File: day02/BinarySearch.js
function getRandom(min, max) {
  return Math.floor(Math.random() * (max - min + 1) + min)
}

let arr = []
for (let i = 0; i < 1000; i++) {
  const random = getRandom(0, 1000)
  arr.push(random)
}

arr = arr.sort((a, b) => a - b)

console.log(arr)

// 0,1,2,3,4,5,6,7,8,9

function binarySearch(arr, target) {
  let l = 0
  let r = arr.length - 1

  while (l <= r) {
    let mid = l + Math.floor((r - l) / 2)

    if (arr[mid] > target) {
      r = mid - 1
    } else if (arr[mid] < target) {
      l = mid + 1
    } else {
      return mid
    }
  }

  return -1
}

console.log(binarySearch(arr, 10))

File: day02/BinarySearch.cpp
#include <iostream>
#include <cmath>
#include <cassert>
#include <ctime>

using namespace std;

template <typename T>
int binarySearch(T arr[], int n, T target)
{
  int l = 0, r = n - 1; // 在[l...r]的范围里寻找target
  while (l <= n)        // 当l==r时，区间[l...r]依然是有效的
  {
    int mid = l + (r - l) / 2;

    if (arr[mid] > target)
    {
      r = mid - 1; // target在[l,mid-1]中
    }
    else if (arr[mid] < target)
    {
      l = mid + 1; // target在[mid+1...r]中
    }
    else
    {
      return mid;
    }
  }

  return -1;
}

int main()
{
  int arr[] = {1, 2, 3, 4, 5, 6, 7, 8, 9, 10};
  int length = 10;
  int result = binarySearch(arr, length, 2);
  cout << result << endl;
  return 0;
}